scripts/rollup/build.js

```diff
@@ -393,61 +409,39 @@ function getPlugins(
         !isProduction
       )
     ),
-    // Remove 'use strict' from individual source files.
-    {
-      transform(source) {
-        return source.replace(/['"]use strict["']/g, '');
-      },
-    },
-    // Turn __DEV__ and process.env checks into constants.
-    replace({
-      __DEV__: isProduction ? 'false' : 'true',
-      __PROFILE__: isProfiling || !isProduction ? 'true' : 'false',
-      __UMD__: isUMDBundle ? 'true' : 'false',
-      'process.env.NODE_ENV': isProduction ? "'production'" : "'development'",
-      __EXPERIMENTAL__,
-      // Enable forked reconciler.
-      // NOTE: I did not put much thought into how to configure this.
-      __VARIANT__: bundle.enableNewReconciler === true,
-    }),
     // The CommonJS plugin *only* exists to pull "art" into "react-art".
     // I'm going to port "art" to ES modules to avoid this problem.
     // Please don't enable this for anything else!
     isUMDBundle && entry === 'react-art' && commonjs(),
+
     // Apply dead code elimination and/or minification.
     isProduction &&
-      closure(
-        Object.assign({}, closureOptions, {
-          // Don't let it create global variables in the browser.
-          // https://github.com/facebook/react/issues/10909
-          assume_function_wrapper: !isUMDBundle,
-          renaming: !shouldStayReadable,
-        })
-      ),
-    // HACK to work around the fact that Rollup isn't removing unused, pure-module imports.
-    // Note that this plugin must be called after closure applies DCE.
-    isProduction && stripUnusedImports(pureExternalModules),
-    // Add the whitespace back if necessary.
+      terser({
+        // Don't let it create global variables in the browser.
+        enclose: !isUMDBundle,
+        mangle: !shouldStayReadable,
+      }),
     shouldStayReadable &&
       prettier({
         parser: 'babel',
         singleQuote: false,
         trailingComma: 'none',
         bracketSpacing: true,
+        sourceMap: false,
       }),
-    // License and haste headers, top-level `if` blocks.
-    {
-      renderChunk(source) {
-        return Wrappers.wrapBundle(
-          source,
-          bundleType,
-          globalName,
-          filename,
-          moduleType
-        );
-      },
-    },
-    // Record bundle size.
+    // // License and haste headers, top-level `if` blocks.
+    // {
+    //   renderChunk(source) {
+    //     return Wrappers.wrapBundle(
+    //       source,
+    //       bundleType,
+    //       globalName,
+    //       filename,
+    //       moduleType
+    //     );
+    //   },
+    // },
```

Comment:
Looks like I missed the bit of code at the end of `wrapBundle`:

```js
  if (moduleType === RECONCILER) {
    // Standalone reconciler is only used by third-party renderers.
    // It is handled separately.
    const wrapper = reconcilerWrappers[bundleType];
    if (typeof wrapper !== 'function') {
      throw new Error(
        `Unsupported build type for the reconciler package: ${bundleType}.`
      );
    }
    return wrapper(source, globalName, filename, moduleType);
  }
  // All the other packages.
  const wrapper = wrappers[bundleType];
  if (typeof wrapper !== 'function') {
    throw new Error(`Unsupported build type: ${bundleType}.`);
  }
  return wrapper(source, globalName, filename, moduleType);
```

Each of these functions (or the code that calls them) would need to similarly make use of `magic-string` to not break the sourcemap support:

- `wrappers`: https://github.com/facebook/react/blob/main/scripts/rollup/wrappers.js#L53-L299
- `reconcilerWrappers`: https://github.com/facebook/react/blob/main/scripts/rollup/wrappers.js#L301-L351

From a quick skim through:

- it looks like most of the `wrappers` could use `magic-string`'s `prepend`, though some (eg the UMD_*, NODE_DEV, FB_WWW_DEV, RN_OSS_DEV, RN_FB_DEV, etc builds) would also need to be able to use `append`.
- the `reconcilerWrappers` would all need to use both `prepend` and `append`

While the original code just manipulates the `source` variable throughout this entire function, we would need to ensure we're only manipulating the `MagicString` object with it's functions until the very end, where we can use `s.toString()` or similar in our final `return`.

---

Looking at Rollup's `renderChunk`:

- https://rollupjs.org/guide/en/#renderchunk
  - `(code: string, chunk: ChunkInfo, options: OutputOptions) => string | { code: string, map: SourceMap } | null`
- https://rollupjs.org/guide/en/#source-code-transformations
  - > If a plugin transforms source code, it should generate a sourcemap automatically, unless there's a specific `sourceMap: false` option. Rollup only cares about the `mappings` property (everything else is handled automatically). If it doesn't make sense to generate a sourcemap, (e.g. [rollup-plugin-string](https://github.com/TrySound/rollup-plugin-string)), return an empty sourcemap

It looks like to enable sourcemap support we would need to return an object `{ code: string, map: SourceMap }` rather than just the current `string`.

We can use `magic-string`'s `s.generateMap( options )`:

- https://github.com/rich-harris/magic-string#sgeneratemap-options-

Which would end up looking something like:

```js
const s = new MagicString(source);

// do some manipulations with s.prepend, s.append, s.replace, etc

// TODO: maybe add some other options to s.generateMap as well?
return { code: s.toString(), map: s.generateMap({ source: filename }) };
```

Reply:
It looks like `magic-string` only added `s.replace` in `0.26.0`, which requires Node.js v12 or higher:

- https://github.com/Rich-Harris/magic-string/blob/master/CHANGELOG.md#0260-2022-03-03

The current `yarn.lock` contains the following versions:

- `magic-string@0.25.6`
- `magic-string@0.25.7`

```
⇒  yarn why magic-string
yarn why v1.22.19
[1/4] 🤔  Why do we have the module "magic-string"...?
[2/4] 🚚  Initialising dependency graph...
[3/4] 🔍  Finding dependency...
[4/4] 🚡  Calculating file sizes...
=> Found "magic-string@0.25.7"
info Has been hoisted to "magic-string"
info Reasons this module exists
   - "workspace-aggregator-0c762641-e299-4a7b-b334-fee60ce16010" depends on it
   - Hoisted from "_project_#rollup-plugin-prettier#magic-string"
   - Hoisted from "_project_#rollup-plugin-strip-banner#magic-string"
   - Hoisted from "_project_#@rollup#plugin-replace#magic-string"
   - Hoisted from "_project_#@rollup#plugin-commonjs#magic-string"
info Disk size without dependencies: "388KB"
info Disk size with unique dependencies: "456KB"
info Disk size with transitive dependencies: "456KB"
info Number of shared dependencies: 1
=> Found "rollup-plugin-commonjs#magic-string@0.25.6"
info This module exists because "_project_#react-devtools-extensions#rollup-plugin-commonjs" depends on it.
✨  Done in 1.08s.
```

Looking at [`.circleci/config.yml`](https://github.com/facebook/react/blob/main/.circleci/config.yml#L5), the jobs seem to run on the `cimg/openjdk:17.0.0-node` docker container: 

- https://circleci.com/developer/images/image/cimg/openjdk
	- `17.0.0-node` uses: `node 14.17.6`, `yarn 1.22.5`

But looking at [`appveyor.yml`](https://github.com/facebook/react/blob/main/appveyor.yml#L10), Windows seems to be tested against node `10.x`, which would presumably fail to build if we updated to `magic-string` `0.26.0` or higher, since it relies on Node 12.x or higher.

This PR attempted to add testing support to AppVeyor for Node 11.x and 12.x, but never landed:

- https://github.com/facebook/react/pull/17197

Reply:
It looks like the potential concerns raised around `appveyor.yml` and node versions is now resolved (the config was unused, and has been removed), as per the following:

> Based on the following comment on the associated PR I opened, I believe this issue can now be closed as well:
> 
> > The unused config was now deleted. Thanks for drawing attention at it!
> > 
> > _Originally posted by @kassens in https://github.com/facebook/react/issues/24892#issuecomment-1265459348_
> 
> _Originally posted by @0xdevalias in https://github.com/facebook/react/issues/24891#issuecomment-1487828154_